extension-manifest-v3/src/background/adblocker.js

```diff
@@ -24,102 +25,116 @@ import { getMetadata } from '/utils/trackerdb.js';
-    engines.setEnv('env_experimental', value);
+    console.info(
+      `[adblocker] engine reloaded with: ${enabledEngines.join(', ')}`,
+    );
+  } else {
+    engines.create(engines.MAIN_ENGINE);
+    console.info('[adblocker] Engine reloaded with no filters');
+  }
+}
 
-    // As engines on the server might have new filters, we force the update
-    // when value has changed from false to true
-    if (lastValue !== undefined && value) {
-      engines.updateAll().catch(() => null);
-    }
-  }),
+engines.addChangeListener(engines.CUSTOM_ENGINE, async () => {
+  reloadEngine(getEnabledEngines(options));
+});
 
-  // Regional filters
-  observe('regionalFilters', async ({ enabled, regions }, lastValue) => {
-    // Background startup
-    if (!lastValue) {
-      const engine = await regionalFiltersEngine;
-      // Pre-requirement for skipping update - engine must be initialized
-      // Otherwise it is a very first try to setup the engine
-      if (engine.lists.size) return;
-    }
+export const setup = asyncSetup([
+  observe(async (value, lastValue) => {
+    options = value;
 
-    // Clean previous regional engines
-    if (lastValue) {
-      lastValue.regions
-        .filter((id) => !regions.includes(id))
-        .forEach((id) => engines.clean(`lang-${id}`));
+    const enabledEngines = getEnabledEngines(value);
+    const prevEnabledEngines = lastValue && getEnabledEngines(lastValue);
+
+    if (
+      // Reload/mismatched main engine
+      !(await engines.init(engines.MAIN_ENGINE)) ||
+      // Enabled engines changed
+      (prevEnabledEngines &&
+        (enabledEngines.length !== prevEnabledEngines.length ||
+          enabledEngines.some((id, i) => id !== prevEnabledEngines[i]))) ||
+      // Engine filters updatedAt changed after successful update
+      (lastValue && value.filtersUpdatedAt > lastValue.filtersUpdatedAt)
+    ) {
+      // The regional filters engine is no longer used, so we must remove it
+      // from the storage. We do it as rarely as possible, to avoid unnecessary loads.
+      // TODO: this can be removed in the future release when most of the users will have
+      // the new version of the extension
+      engines.remove('regional-filters');
+
+      await reloadEngine(enabledEngines);
     }
 
-    // Schedule merge when one of the regional engines is updated
-    regions.forEach((id) => {
-      engines.addUpdateListener(`lang-${id}`, mergeRegionalEngines);
-    });
+    if (
+      // Experimental filters enabled
+      (options.experimentalFilters &&
+        lastValue?.experimentalFilters === false) ||
+      // Filters outdated (1 hour)
+      options.filtersUpdatedAt < Date.now() - HOUR_IN_MS
+    ) {
+      const updateEngines = getEnabledEngines(options)
+        // Remove the custom engine, as it is created client-side only
+        .filter((name) => name !== engines.CUSTOM_ENGINE)
+        // Add engines outside of the main engine
+        .concat(engines.TRACKERDB_ENGINE);
```

Comment:
```suggestion
      const updateEngines = new Set([
        ...getEnabledEngines(options),
        engines.TRACKERDB_ENGINE,
      ]);
      updateEngines.delete(engines.CUSTOM_ENGINE);
```

Reply:
Apple vs oranges ;) You create an array on the fly, trigger the spread operator (so iteration over all of the enabled engines), then also remove one of them.. Looks less readable, and I doubt that it's more preformant. This action is taken at most once an hour (or when user forces it). I recommend not to make premature optimizations.